src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -1345,8 +1346,34 @@ Pipe ReadFromMergeTree::spreadMarkRangesAmongStreamsFinal(
         /// can use parallel select on such parts.
         bool no_merging_final = do_not_merge_across_partitions_select_final &&
             std::distance(parts_to_merge_ranges[range_index], parts_to_merge_ranges[range_index + 1]) == 1 &&
-            parts_to_merge_ranges[range_index]->data_part->info.level > 0 &&
-            data.merging_params.is_deleted_column.empty() && !reader_settings.read_in_order;
+            parts_to_merge_ranges[range_index]->data_part->info.level > 0 && !reader_settings.read_in_order;
+
+        // if do_not_merge_across_partitions_select_final = 1 and the partition contains a single part,
+        // there is no reason to handle `is_deleted` using slow inorder reading. Faster to do it with prewhere
+        if (no_merging_final && !data.merging_params.is_deleted_column.empty())
+        {
+            auto equal_function = FunctionFactory::instance().get("equals", context);
+
+            if (!prewhere_info)
+            {
+                prewhere_info = std::make_shared<PrewhereInfo>();
+            }
+
+            const auto * constant_node = &prewhere_info->prewhere_actions.addColumn(
+                ColumnWithTypeAndName{
+                    ColumnUInt8::create(0),
+                    std::make_shared<DataTypeUInt8>(),
+                    ""
+                });
+
+            prewhere_info->prewhere_column_name = data.merging_params.is_deleted_column;
```

Comment:
This impl works on query with no prewhere, but doesn't for queries that already have a prewhere set. Need to investigate how to use this class properly.